File: grind75/week2/030-Valid-Sudoku-Medium/030-Valid-Sudoku-Medium.cpp
// https://leetcode.com/problems/valid-sudoku

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        // Grid is 9x9

        // Maps will use set since values cannot repeat
        // map => row : values
        // map => col : values
        // map => (r/3, c/3) : values, e.g. what values are in each sub grid
        unordered_map<int, unordered_set<char>> rowMap;
        unordered_map<int, unordered_set<char>> colMap;
        map<pair<int, int>, unordered_set<char>> subGridMap;

        for (int r = 0; r < 9; r++) 
        {
            for (int c = 0; c < 9; c++) 
            {
                if (board[r][c] == '.') 
                {
                  continue;
                }
                else 
                {
                    if (rowMap[r].count(board[r][c]) ||
                        colMap[c].count(board[r][c]) ||
                        subGridMap[{r / 3, c / 3}].count(board[r][c])) 
                      {
                      return false;
                      }
                    

                    rowMap[r].insert(board[r][c]);
                    colMap[c].insert(board[r][c]);
                    subGridMap[{r / 3, c / 3}].insert(board[r][c]);
                }
            } // end of inner for-loop
        } // end of outer for-loop

        return true;
    }
};

File: grind75/week2/030-Valid-Sudoku-Medium/030-Valid-Sudoku-Medium.py
# https://leetcode.com/problems/valid-sudoku

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # Grid is 9x9

        # Maps will use set since values cannot repeat
        # map => row : values
        # map => col : values
        # map => (r/3, c/3) : values, e.g what values are in each sub grid
        rowMap = defaultdict(set)
        colMap = defaultdict(set)
        subGridMap = defaultdict(set)
        for r in range(9):
            for c in range(9):
                if board[r][c] == ".":
                    continue
                else:
                    if (board[r][c] in rowMap[r] or
                       board[r][c] in colMap[c] or
                       board[r][c] in subGridMap[(r//3, c//3)]):
                        return False
                    rowMap[r].add(board[r][c])
                    colMap[c].add(board[r][c])
                    subGridMap[(r//3, c//3)].add(board[r][c])
        return True
                    

                    
